### ISSUE TITLE
KSQLDB: JOINs on multiple conditions are not yet supported

### ISSUE THREAD

**Issue by NCAndreea** [confluentinc/ksql#8574]
Hi,

I have following scenario in which I need to join multiple tables to store the result in another Kafka topic.
Unfortunately looks like KSQLDB does not support join condition to be formed on multiple columns.
Moreover if I remove the join that caused this error I get another error because the tables primary keys are different.
Please can you advise what option/solution do I have?
Thank you for your help.

Regards,

How to reproduce
CREATE TABLE TEST1 (ID VARCHAR PRIMARY KEY, Country_ID VARCHAR, Code1 VARCHAR)
WITH (kafka_topic='test1', value_format='JSON', partitions=3);

INSERT INTO TEST1 (ID, Country_ID, Code1) VALUES ('1','1','A');

CREATE TABLE TEST2 (Country_ID VARCHAR PRIMARY KEY, Country_Name VARCHAR)
WITH (kafka_topic='test2', value_format='JSON', partitions=3);

INSERT INTO TEST2 (Country_ID, Country_Name) VALUES ('1','Spain');

CREATE TABLE TEST3 (Country_ID VARCHAR PRIMARY KEY, Code1 VARCHAR PRIMARY KEY, Code_description VARCHAR)
WITH (kafka_topic='test3', value_format='JSON', key_format='JSON', partitions=3);

INSERT INTO TEST3 (Country_ID, Code1, Code_description) VALUES ('1','A', 'ABC');

select
TEST1.ID,
TEST2.Country_Name,
TEST3.Code_description
from TEST1 left join TEST2 on TEST1.Country_ID=TEST2.Country_ID
left join TEST3 on TEST1.Country_ID=TEST3.Country_ID and TEST1.Code1=TEST3.Code1
emit changes limit 5;

![image](https://user-images.githubusercontent.com/96262788/148922053-42572603-bc47-44e6-a726-e4a109578e9b.png)

**Comment by OneCricketeer** [comment]
Related https://github.com/confluentinc/ksql/issues/2344

Also unclear what version you're using, but there are tests for 3-way joins - https://github.com/confluentinc/ksql/pull/7606

**Comment by NCAndreea** [comment]
Yes indeed here I saw the example, but is between a stream and multiple tables and in my case they are all tables.
https://docs.ksqldb.io/en/latest/developer-guide/joins/join-streams-and-tables/#n-way-joins

**Comment by suhas-satish** [comment]
This is a duplicate of https://github.com/confluentinc/ksql/issues/6484
Closing this.

**Comment by NCAndreea** [comment]
I do not think this is same as the one you linked. Please can you check again. 
In that example: 
CREATE STREAM joined AS 
  SELECT * 
  FROM A
    **LEFT JOIN _B_ b1 ON A.b1_id = b1.id
    LEFT JOIN _B_ b2 ON A.b2_id = b2.id;**
    
My example: 
 select
TEST1.ID,
TEST2.Country_Name,
TEST3.Code_description
from TEST1 left join _**TEST2**_ on TEST1.Country_ID=TEST2.Country_ID
left join _**TEST3**_ on **TEST1.Country_ID=TEST3.Country_ID and TEST1.Code1=TEST3.Code1**
emit changes limit 5;

**Comment by vcrfxia** [comment]
Hey all, this is definitely a feature gap in ksqlDB that should be patched sooner rather than later. In the meantime, you can work around this limitation by using a single struct-type key rather than having multiple key columns: https://www.confluent.io/blog/ksqldb-0-15-reads-more-message-keys-supports-more-data-types/#join-on-sources

Specifically, rewrite your source table
```
CREATE TABLE TEST3 (Country_ID VARCHAR PRIMARY KEY, Code1 VARCHAR PRIMARY KEY, Code_description VARCHAR)
WITH (kafka_topic='test3', value_format='JSON', key_format='JSON', partitions=3);
```
as
```
CREATE TABLE TEST3 (comp_key STRUCT<Country_ID VARCHAR, Code1 VARCHAR> PRIMARY KEY, Code_description VARCHAR)
WITH (kafka_topic='test3', value_format='JSON', key_format='JSON', partitions=3);
```
in order to work around the need for a multi-part join condition.

As for adding native support for this in ksqlDB, we should certainly do it. It might also be nice to think through the entirety of what we want our multi-column key support to look like while doing so, including https://github.com/confluentinc/ksql/issues/8489 as well as the other leftover tickets from https://github.com/confluentinc/ksql/projects/3 and https://github.com/confluentinc/ksql/projects/9.

**Comment by MaxPrimeAERY** [comment]
Any suggestion to this question ?

**Comment by woiza** [comment]
Any news on when this will be supported?

**Comment by Taymindis** [comment]
> Hey all, this is definitely a feature gap in ksqlDB that should be patched sooner rather than later. In the meantime, you can work around this limitation by using a single struct-type key rather than having multiple key columns: https://www.confluent.io/blog/ksqldb-0-15-reads-more-message-keys-supports-more-data-types/#join-on-sources
> 
> Specifically, rewrite your source table
> 
> ```
> CREATE TABLE TEST3 (Country_ID VARCHAR PRIMARY KEY, Code1 VARCHAR PRIMARY KEY, Code_description VARCHAR)
> WITH (kafka_topic='test3', value_format='JSON', key_format='JSON', partitions=3);
> ```
> 
> as
> 
> ```
> CREATE TABLE TEST3 (comp_key STRUCT<Country_ID VARCHAR, Code1 VARCHAR> PRIMARY KEY, Code_description VARCHAR)
> WITH (kafka_topic='test3', value_format='JSON', key_format='JSON', partitions=3);
> ```
> 
> in order to work around the need for a multi-part join condition.
> 
> As for adding native support for this in ksqlDB, we should certainly do it. It might also be nice to think through the entirety of what we want our multi-column key support to look like while doing so, including #8489 as well as the other leftover tickets from https://github.com/confluentinc/ksql/projects/3 and https://github.com/confluentinc/ksql/projects/9.

Hi, The sad news is group by struct key is not working for AVRO and PROTOBUF, only working for JSON. 


[Case by case reference](https://stackoverflow.com/questions/78922617/schema-registry-409-when-issuing-pull-query-against-table-with-struct-grouped)

The surprisingly happened was select in condition with more than 1 input are working fine.
